source/extensions/common/redis/redirection_mgr_impl.h

```diff
@@ -60,13 +66,29 @@ class RedirectionManagerImpl : public RedirectionManager,
       : main_thread_dispatcher_(main_thread_dispatcher), cm_(cm), time_source_(time_source) {}
 
   bool onRedirection(const std::string& cluster_name) override;
+  bool onFailure(const std::string& cluster_name) override;
+  bool onHostDegraded(const std::string& cluster_name) override;
 
   HandlePtr registerCluster(const std::string& cluster_name,
                             std::chrono::milliseconds min_time_between_triggering,
-                            uint32_t redirects_threshold, const RedirectCB& cb) override;
+                            const uint32_t redirects_threshold, const uint32_t failure_threshold,
+                            const uint32_t host_degraded_threshold, const RedirectCB& cb) override;
 
 private:
   void unregisterCluster(const ClusterInfoSharedPtr& cluster_info);
+  /**
+   * The type of events that can trigger discovery
+   */
+  enum EventType {
+    // MOVE or ASK redirection
+    Redirection,
+    // Failure
+    Failure,
+    // Sending request to degraded/unhealthy host
+    DegradedHost
+  };
+
+  bool onEvent(const std::string& cluster_name, EventType event_type);
```

Comment:
This is now doing more than being a redirection manager. Can you rename this something like TopologyRefreshManager or something like that? (Renames files also)